src/os_version_check/darwin_impl.rs

```diff
@@ -0,0 +1,590 @@
+use core::ffi::CStr;
+use core::{
+    ffi::{c_char, c_int, c_long, c_uint, c_void},
+    num::NonZero,
+    ptr::null_mut,
+    slice,
+    sync::atomic::{AtomicU32, Ordering},
+};
+
+/// Get the current OS version.
+#[inline]
+pub(super) fn current_version() -> u32 {
+    // Cache the lookup for performance.
+    //
+    // 0.0.0 is never gonna be a valid version, so we use that as our sentinel value.
+    static CURRENT_VERSION: AtomicU32 = AtomicU32::new(0);
+
+    // We use relaxed atomics, it doesn't matter if multiple threads end up racing to read or write
+    // the version, `lookup_version` should be idempotent and always return the same value.
+    //
+    // `compiler-rt` uses `dispatch_once`, but that's overkill for the reasons above.
+    let version = CURRENT_VERSION.load(Ordering::Relaxed);
+    if version == 0 {
+        let version = lookup_version().get();
+        CURRENT_VERSION.store(version, Ordering::Relaxed);
+        version
+    } else {
+        version
+    }
+}
+
+#[cold]
+// Use `extern "C"` to abort on panic, allowing `current_version` to be free of panic handling.
+pub(super) extern "C" fn lookup_version() -> NonZero<OSVersion> {
+    // Since macOS 10.15, libSystem has provided the undocumented `_availability_version_check` via
+    // `libxpc` (zippered, so requires platform parameter to differentiate between on macOS and Mac
+    // Catalyst) for doing the version lookup, though it's usage may be a bit dangerous, see:
+    // - https://reviews.llvm.org/D150397
+    // - https://github.com/llvm/llvm-project/issues/64227
+    //
+    // So instead, we use the safer approach of reading from `sysctl` (which is faster), and if that
+    // fails, we fall back to the property list (this is what `_availability_version_check` does
+    // internally).
+    let version = version_from_sysctl().unwrap_or_else(version_from_plist);
+
+    // Use `NonZero` to try to make it clearer to the optimizer that this will never return 0.
+    NonZero::new(version).expect("version cannot be 0.0.0")
+}
+
+/// Look up the current OS version(s) from `/System/Library/CoreServices/SystemVersion.plist`.
+///
+/// More specifically, from the `ProductVersion` and `iOSSupportVersion` keys, and from
+/// `$IPHONE_SIMULATOR_ROOT/System/Library/CoreServices/SystemVersion.plist` on the simulator.
+///
+/// This file was introduced in macOS 10.3, which is well below the minimum supported version by
+/// `rustc`, which is currently macOS 10.12.
+///
+/// # Panics
+///
+/// Panics if reading or parsing the version fails (or if the system was out of memory).
+///
+/// We deliberately choose to panic, as having this silently return an invalid OS version would be
+/// impossible for a user to debug.
```

Comment:
compiler-builtins may not panic as that requires libcore but compiler-builtins can't depend on any other crates as it is last on the liker cli. Instead you will have to abort.

Reply:
> as it is last on the liker cli

Pretty sure Apple's linker doesn't require ordering in the same way as traditional ELF linkers?

I'll have to test it to be sure though, I'll see if I can build a custom toolchain with this branch of `compiler-builtins`.

Reply:
Even if the linker accepts it, you run the risk of cyclic dependencies with the panic handler (as `__isPlatformVersionAtLeast` panics, which calls the panic handler, which in turn calls a function that calls `__isPlatformVersionAtLeast`).